### PR TITLE
Opportunistically use dense iter for archetypal iteration in Par_iter

### PR DESCRIPTION
**Pull request by re0312**
# Objective

- follow of #14049 ,we could use it on our Parallel Iterator,this pr also unified  the used function in both regular iter and parallel iterations.

## Performance 

image

no performance regression for regular itertaion

3.5X faster in hybrid parallel iteraion,this number is far greater than the benefits obtained in regular iteration(~1.81) because mutable iterations on continuous memory can effectively reduce the cost of mataining core cache coherence

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
Dang, those are some *impressive* numbers.

**Review by chescock — APPROVED** [review]

**Review by NthTensor — APPROVED** [review]
I don't feel qualified to approve the code but I did replicate the results.

File: crates/bevy_ecs/src/query/iter.rs [hunk 1]
```diff
@@ -122,6 +122,66 @@ impl<'w, 's, D: QueryData, F: QueryFilter> QueryIter<'w, 's, D, F> {
         }
     }
 
+    /// Executes the equivalent of [`Iterator::fold`] over a contiguous segment
+    /// from an storage.
+    ///
+    ///  - `range` must be in `[0, storage::entity_count)` or None.
```
Comment (13ros27) [inline crates/bevy_ecs/src/query/iter.rs:128]: This should probably be formatted like public facing safety docs with `/// # Safety` above it. As an aside for a later time we should consider enabling `clippy::check-private-items` to catch these.

**Review by 13ros27 — APPROVED** [review]
This looks good to me, I didn't run the benchmarks or scrutinise them very much but I trust your and Nths numbers. The comments are very clear and easy to follow so thank you for those, nice bit of deduplication too.

**Review by alice-i-cecile — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1682 if you'd like to help out.

**Pull request merged**